# PHASE 1
# Xử lí dữ liệu

In [25]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

---
## Giới thiệu dataset

Dữ liệu giao dịch trong thời gian **01/01/2019 — 31/12/2019**, gồm 5 file:

| File | Mô tả | Ghi chú |
|---|---|---|
| `Online_Sales.csv` | Dữ liệu giao dịch ở cấp độ từng sản phẩm | File chính |
| `CustomersData.xlsx` | Thông tin khách hàng khách hàng | Giới tính, địa điểm, tenure |
| `Discount_Coupon.csv` | Mã giảm giá theo tháng & danh mục sản phẩm | liên kết theo (Month vs Category) |
| `Tax_amount.xlsx` | Tỷ lệ GST theo danh mục | liên kết theo Category |
| `Marketing_Spend.csv` | Chi phí marketing offline & online theo ngày | liên kết theo date |

### Mô tả cột từng bảng

**Online_Sales** (bảng chính):

| Cột | Kiểu | Mô tả |
|---|---|---|
| CustomerID | int | ID khách hàng |
| Transaction_ID | int | ID đơn hàng — **không unique** (1 đơn có nhiều sản phẩm) |
| Transaction_Date | str | Ngày giao dịch (dd/mm/yyyy) |
| Product_SKU | str | Mã sản phẩm |
| Product_Description | str | Tên - mô tả sản phẩm |
| Product_Category | str | Danh mục sản phẩm — liên kết với bảng Tax_amount |
| Quantity | int | Số lượng sản phẩm được mua |
| Avg_Price | float | Đơn giá |
| Delivery_Charges | float | Phí giao hàng |
| Coupon_Status | str | `Used` / `Clicked` / `Not Used` |

**CustomersData:**

| Cột | Kiểu | Mô tả |
|---|---|---|
| CustomerID | int | ID khách hàng - liên kết với bảng Online_Sales(CustomerID) |
| Gender | str | Giới tính (M/F) |
| Location | str | Thành phố/khu vực |
| Tenure_Months | int | Số tháng là khách hàng |

**Discount_Coupon:**

| Cột | Kiểu | Mô tả |
|---|---|---|
| Month | str | Tháng áp dụng (Jan, Feb, ...) |
| Product_Category | str | Danh mục áp dụng |
| Coupon_Code | str | Mã coupon |
| Discount_pct | int | % giảm giá (10, 20, 30...) |

**Tax_amount:**

| Cột | Kiểu | Mô tả |
|---|---|---|
| Product_Category | str | liên kết với bảng Online_Sales(Product_Category) |
| GST | float | Tỷ lệ thuế (0.05, 0.10, 0.18...) |

**Marketing_Spend:**

| Cột | Kiểu | Mô tả |
|---|---|---|
| Date | str | Ngày thực hiện Marketing - liên kết với bảng Online_Sales(Transaction_Date) |
| Offline_Spend | int | Chi phí TV, radio, báo, billboard |
| Online_Spend | float | Chi phí Google Ads, Facebook... |


---
## Load data

In [27]:
sales = pd.read_csv("data/Online_Sales.csv")
cust  = pd.read_excel("data/CustomersData.xlsx")
disc  = pd.read_csv("data/Discount_Coupon.csv")
tax   = pd.read_excel("data/Tax_amount.xlsx")
mkt   = pd.read_csv("data/Marketing_Spend.csv")
 
print(f"Online_Sales (sales)    : {sales.shape[0]:,} rows × {sales.shape[1]} cols")
print(f"CustomersData (cust)    : {cust.shape[0]:,} rows × {cust.shape[1]} cols")
print(f"Discount_Coupon (disc)  : {disc.shape[0]:,} rows × {disc.shape[1]} cols")
print(f"Tax_amount (tax)        : {tax.shape[0]:,} rows × {tax.shape[1]} cols")
print(f"Marketing_Spend (mkt)   : {mkt.shape[0]:,} rows × {mkt.shape[1]} cols")

Online_Sales (sales)    : 52,924 rows × 10 cols
CustomersData (cust)    : 1,468 rows × 4 cols
Discount_Coupon (disc)  : 204 rows × 4 cols
Tax_amount (tax)        : 20 rows × 2 cols
Marketing_Spend (mkt)   : 365 rows × 3 cols


In [3]:
# Kiểm tra số dòng số cột, dữ liệu null, duplicated cột của từng bảng
for name, df in [("Online_Sales", sales), ("CustomersData", cust),
                 ("Discount_Coupon", disc), ("Tax_amount", tax),
                 ("Marketing_Spend", mkt)]:
    print(f"\n[{name}]")
    print(f"  Shape      : {df.shape}")
    print(f"  Null cells : {df.isnull().sum().sum()}")
    print(f"  Dup rows   : {df.duplicated().sum()}")


[Online_Sales]
  Shape      : (52924, 10)
  Null cells : 0
  Dup rows   : 0

[CustomersData]
  Shape      : (1468, 4)
  Null cells : 0
  Dup rows   : 0

[Discount_Coupon]
  Shape      : (204, 4)
  Null cells : 0
  Dup rows   : 0

[Tax_amount]
  Shape      : (20, 2)
  Null cells : 0
  Dup rows   : 0

[Marketing_Spend]
  Shape      : (365, 3)
  Null cells : 0
  Dup rows   : 0


---
## Khám phá từng bảng

### 1. Online_Sales

In [4]:
display(sales.head(3))
print(sales.info())
display("\n",sales.describe().round(2))

,CustomerID,Transaction_ID,Transaction_Date,Product_SKU,Product_Description,Product_Category,Quantity,Avg_Price,Delivery_Charges,Coupon_Status
0,17850,16679,1/1/2019,GGOENEBJ079499,Nest Learning Thermostat 3rd Gen-USA - Stainle...,Nest-USA,1,153.71,6.5,Used
1,17850,16680,1/1/2019,GGOENEBJ079499,Nest Learning Thermostat 3rd Gen-USA - Stainle...,Nest-USA,1,153.71,6.5,Used
2,17850,16681,1/1/2019,GGOEGFKQ020399,Google Laptop and Cell Phone Stickers,Office,1,2.05,6.5,Used


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52924 entries, 0 to 52923
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   CustomerID           52924 non-null  int64  
 1   Transaction_ID       52924 non-null  int64  
 2   Transaction_Date     52924 non-null  object 
 3   Product_SKU          52924 non-null  object 
 4   Product_Description  52924 non-null  object 
 5   Product_Category     52924 non-null  object 
 6   Quantity             52924 non-null  int64  
 7   Avg_Price            52924 non-null  float64
 8   Delivery_Charges     52924 non-null  float64
 9   Coupon_Status        52924 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 4.0+ MB
None


'\n'

,CustomerID,Transaction_ID,Quantity,Avg_Price,Delivery_Charges
count,52924.00,52924.00,52924.0,52924.00,52924.00
mean,15346.71,32409.83,4.5,52.24,10.52
std,1766.56,8648.67,20.1,64.01,19.48
min,12346.00,16679.00,1.0,0.39,0.00
25%,13869.00,25384.00,1.0,5.70,6.00
50%,15311.00,32625.50,1.0,16.99,6.00
75%,16996.25,39126.25,2.0,102.13,6.50
max,18283.00,48497.00,900.0,355.74,521.36


### 2. CustomersData

In [5]:
display(cust.head(3))
print(cust.info())
print("\n",cust.describe().round(2))

# Kiểm tra CustomerID của 2 bảng Online_Sales và CustomersData có khớp nhau không
check = set(sales['CustomerID'].unique()) - set(cust['CustomerID'].unique())
print(f"\n[Join check] CustomerID khớp" if len(check)==0
      else f"\n {len(check)} CustomerID không có trong CustomersData")

# Phân phối của Gender và Location
print("\n""Phân phối dữ liệu", cust['Gender'].value_counts().to_string())
print("\n""Phân phối dữ liệu", cust['Location'].value_counts().to_string())

,CustomerID,Gender,Location,Tenure_Months
0,17850,M,Chicago,12
1,13047,M,California,43
2,12583,M,Chicago,33


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1468 entries, 0 to 1467
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   CustomerID     1468 non-null   int64 
 1   Gender         1468 non-null   object
 2   Location       1468 non-null   object
 3   Tenure_Months  1468 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 46.0+ KB
None

        CustomerID  Tenure_Months
count     1468.00        1468.00
mean     15314.39          25.91
std       1744.00          13.96
min      12346.00           2.00
25%      13830.50          14.00
50%      15300.00          26.00
75%      16882.25          38.00
max      18283.00          50.00

[Join check] CustomerID khớp

Phân phối dữ liệu Gender
F    934
M    534

Phân phối dữ liệu Location
California       464
Chicago          456
New York         324
New Jersey       149
Washington DC     75


### 3. Discount_Coupon

In [6]:
display(disc.head(3))
display(disc.describe().round())

# Kiểm tra mỗi tháng có bao nhiêu danh mục riêng biệt được giảm giá
print("\n","Số Product_Category được discount mỗi tháng:")
print(disc.groupby('Month')['Product_Category'].count())

# Kiểm tra Product_Category của 2 bảng Online_Sales và Discount_Coupon có khớp nhau không
sales_cats = set(sales['Product_Category'].unique())
disc_cats  = set(disc['Product_Category'].unique())
print(f"\n[Category] Sales không có trong Discount: {sales_cats - disc_cats}")
print(f"[Category] Discount không có trong Sales : {disc_cats - sales_cats}")

,Month,Product_Category,Coupon_Code,Discount_pct
0,Jan,Apparel,SALE10,10
1,Feb,Apparel,SALE20,20
2,Mar,Apparel,SALE30,30


,Discount_pct
count,204.0
mean,20.0
std,8.0
min,10.0
25%,10.0
50%,20.0
75%,30.0
max,30.0



 Số Product_Category được discount mỗi tháng:
Month
Apr    17
Aug    17
Dec    17
Feb    17
Jan    17
Jul    17
Jun    17
Mar    17
May    17
Nov    17
Oct    17
Sep    17
Name: Product_Category, dtype: int64

[Category] Sales không có trong Discount: {'Fun', 'More Bags', 'Backpacks', 'Google'}
[Category] Discount không có trong Sales : {'Notebooks'}


### 4. Tax_amount

In [7]:
display(tax.head())
print(tax.info())
display(tax.describe().round(2))

# Kiểm tra Product_Category của 2 bảng Tax_amount và Online_Sales có khớp nhau không
sales_cats = set(sales['Product_Category'].unique())
tax_cats  = set(tax['Product_Category'].unique())
print(f"\n[Category] Sales không có trong Tax: {sales_cats - tax_cats}")
print(f"[Category] tax không có trong Sales : {tax_cats - sales_cats}")

,Product_Category,GST
0,Nest-USA,0.10
1,Office,0.10
2,Apparel,0.18
3,Bags,0.18
4,Drinkware,0.18


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Product_Category  20 non-null     object 
 1   GST               20 non-null     float64
dtypes: float64(1), object(1)
memory usage: 448.0+ bytes
None


,GST
count,20.00
mean,0.12
std,0.05
min,0.05
25%,0.09
50%,0.10
75%,0.18
max,0.18



[Category] Sales không có trong Tax: set()
[Category] tax không có trong Sales : set()


### 5. Marketing_Spend

In [8]:
display(mkt.head(3))
print(mkt.info())
display(mkt.describe().round())

# Kiểm tra date trong 2 bảng có trùng khớp nhau không
sales_date = set(sales['Transaction_Date'].unique())
mkt_date = set(mkt['Date'].unique())
print(f"\n[Date] Sales không có trong mkt: {sales_date - mkt_date}")
print(f"[Date] mkt không có trong Sales : {mkt_date - sales_date}")

,Date,Offline_Spend,Online_Spend
0,1/1/2019,4500,2424.50
1,1/2/2019,4500,3480.36
2,1/3/2019,4500,1576.38


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           365 non-null    object 
 1   Offline_Spend  365 non-null    int64  
 2   Online_Spend   365 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 8.7+ KB
None


,Offline_Spend,Online_Spend
count,365.0,365.0
mean,2844.0,1906.0
std,952.0,809.0
min,500.0,320.0
25%,2500.0,1259.0
50%,3000.0,1882.0
75%,3500.0,2435.0
max,5000.0,4557.0



[Date] Sales không có trong mkt: set()
[Date] mkt không có trong Sales : set()


---
## Nhận xét

- **Online_Sales:** `Transaction_Date` đang ở dạng object — cần parse sang datetime. `Quantity` có giá trị lên đến 900 nhưng đây là đơn hàng bulk (văn phòng phẩm giá thấp) nên hợp lệ.
- **CustomersData:** Dữ liệu sạch, CustomerID khớp với bảng Online_Sales, không có gì bất thường.
- **Discount_Coupon:** Mỗi tháng có 17 `Product_Category` được discount, có 1 `Product_Category` không có trong online_Sales,  cần gộp 2 cột `Month` và `Product_Category` để liên kết với Online_Sales.
- **Tax_amount:** 20 `Product_Category` khớp với Online_Sales, GST từ 5% đến 18%.
- **Marketing_Spend:** `Date` đang ở dạng object — cần parse sang datetime, cả `date` ở mkt và `Transaction_Date` bảng Online_Sales đêu có đủ 365 ngày trong năm.

---
## Xử lí và làm sạch dữ liệu

In [9]:
# Parse hai cột thời gian từ object sang datetime
sales['Transaction_Date'] = pd.to_datetime(sales['Transaction_Date'])
mkt['Date'] = pd.to_datetime(mkt['Date'])

In [10]:
# Sửa lại tên Product_Category không có trong Online_Sales do nhập liệu sai
disc['Product_Category'] = disc['Product_Category'].replace({'Notebooks': 'Notebooks & Journals'})

In [11]:
# Sau khi sửa lại tên Notebooks bị lỗi nhập liệu thì tiến hành nhóm lại theo 'Month' và 'Product_Category'
# lấy giá trị 'Discount_pct' lớn nhất để tránh nếu trùng nhau
disc_clean = disc.groupby(['Month', 'Product_Category'])['Discount_pct'].max().reset_index()

In [12]:
# Tạo thêm các trường datetime trong bảng Online_Sales để thuận tiện trong việc phân tích
sales['Month'] = sales['Transaction_Date'].dt.strftime('%b')
sales['Year'] = sales['Transaction_Date'].dt.year
sales['Quarter'] = sales['Transaction_Date'].dt.quarter
sales['Week'] = sales['Transaction_Date'].dt.isocalendar().week.astype(int)
sales['DayOfWeek'] = sales['Transaction_Date'].dt.day_name()

---
### Tính Invoice Value
**Invoice Value** = (Quantity × Avg_Price × (1 − Discount_pct) × (1 + GST)) + Delivery_Charges
> **Lưu ý quan trọng:** `Discount_pct` chỉ được áp dụng khi `Coupon_Status = 'Used'`.  
> Hai trạng thái `Clicked` và `Not Used` đều nhận `Discount_pct = 0`.

In [13]:
# Join Discount_Coupon vào Online_Sales
# Dùng left Join để giữ lại toàn bộ giao dịch
# 4 Category không có trong discount sẽ nhận là NaN và sẽ xử lí ở bước sau
df = sales.merge(
    disc_clean[['Month', 'Product_Category', 'Discount_pct']],
    on=['Month', 'Product_Category'],
    how='left')

# Chỉ áp discount khi `Coupon_Status` = 'Used' và đủ điều kiện trong bảng `discount_Coupon`
# Chuyển từ % (10) sang tỉ lệ thập phân (0.10) để tính toán
df['Discount_pct'] = np.where(
    df['Coupon_Status'] == 'Used',
    df['Discount_pct'].fillna(0) / 100,0)

#  Join Tax_amount vào Online_Sales
df = df.merge(
    tax[['Product_Category', 'GST']],
    on='Product_Category',
    how='left')

In [14]:
#  Tính Invoice Value
df['Invoice_Value'] = ((df['Quantity'] * df['Avg_Price'] * (1 - df['Discount_pct']) * (1 + df['GST'])) 
                       + df['Delivery_Charges']).round(2)

# Kiểm tra kết quả (không được có null hoặc giá trị âm)
assert df['Invoice_Value'].isnull().sum() == 0, "Có null trong Invoice_Value"
assert (df['Invoice_Value'] < 0).sum() == 0, "Có giá trị âm trong Invoice_Value"

print("Invoice_Value không có null và không âm")
print(f"Min   : {df['Invoice_Value'].min():.2f}")
print(f"Max   : {df['Invoice_Value'].max():.2f}")
print(f"Mean  : {df['Invoice_Value'].mean():.2f}")
print(f"Total : {df['Invoice_Value'].sum():.2f}")

Invoice_Value không có null và không âm
Min   : 4.60
Max   : 8979.28
Mean  : 101.98
Total : 5397365.02


---
## Xây dựng Star Schema

Sau khi tính Invoice Value, tổ chức lại dữ liệu theo mô hình **Star Schema** gồm 1 fact table và 4 dimension tables — chuẩn bị cho SQL và Power BI

```
dim_customer ←──── fact_sales ────→ dim_product
                        │
                   dim_date
                        │
                   dim_marketing
```

In [15]:
# fact_sales — bảng chính chứa toàn bộ giao dịch
# Chọn các cột cần thiết, bao gồm các cột thời gian đã tạo
# và Invoice_Value vừa tính
fact_sales = df[[
    'Transaction_ID', 'CustomerID', 'Transaction_Date',
    'Year', 'Quarter', 'Month', 'Week', 'DayOfWeek',
    'Product_SKU', 'Product_Description', 'Product_Category',
    'Quantity', 'Avg_Price', 'Delivery_Charges',
    'Coupon_Status', 'Discount_pct', 'GST', 'Invoice_Value']].copy()

In [16]:
# dim_customer (thông tin nhân khẩu học khách hàng)
# Giữ nguyên từ file gốc, chỉ đổi tên rõ ràng hơn
dim_customer = cust.copy()
dim_customer.head()

,CustomerID,Gender,Location,Tenure_Months
0,17850,M,Chicago,12
1,13047,M,California,43
2,12583,M,Chicago,33
3,13748,F,California,30
4,15100,M,California,49


In [17]:
# dim_product — danh mục sản phẩm kèm mức thuế
# Lấy unique SKU từ fact_sales
# để mỗi sản phẩm có đầy đủ thông tin danh mục và thuế suất
dim_product = (
    fact_sales[['Product_SKU', 'Product_Description', 'Product_Category','GST']]
    .drop_duplicates(subset='Product_SKU').reset_index(drop=True).copy())
dim_product.head()

,Product_SKU,Product_Description,Product_Category,GST
0,GGOENEBJ079499,Nest Learning Thermostat 3rd Gen-USA - Stainle...,Nest-USA,0.10
1,GGOEGFKQ020399,Google Laptop and Cell Phone Stickers,Office,0.10
2,GGOEGAAB010516,Google Men's 100% Cotton Short Sleeve Hero Tee...,Apparel,0.18
3,GGOEGBJL013999,Google Canvas Tote Natural/Navy,Bags,0.18
4,GGOEGBMJ013399,Sport Bag,Bags,0.18


In [18]:
# dim_date — calendar table toàn bộ năm 2019
# Tạo bảng ngày đầy đủ 365 ngày để phục vụ time intelligence trong Power BI
# Is_Weekend giúp phân tích hành vi mua sắm cuối tuần vs ngày thường
date_range = pd.date_range('2019-01-01', '2019-12-31', freq='D')
dim_date = pd.DataFrame({
    'Date': date_range,
    'Year': date_range.year,
    'Quarter': date_range.quarter,
    'Month_Num': date_range.month,
    'Month_Name': date_range.strftime('%b'),
    'Week': date_range.isocalendar().week.astype(int),
    'DayOfWeek': date_range.day_name(),
    'Is_Weekend': date_range.dayofweek >= 5})
dim_date.head()

,Date,Year,Quarter,Month_Num,Month_Name,Week,DayOfWeek,Is_Weekend
2019-01-01,2019-01-01,2019,1,1,Jan,1,Tuesday,False
2019-01-02,2019-01-02,2019,1,1,Jan,1,Wednesday,False
2019-01-03,2019-01-03,2019,1,1,Jan,1,Thursday,False
2019-01-04,2019-01-04,2019,1,1,Jan,1,Friday,False
2019-01-05,2019-01-05,2019,1,1,Jan,1,Saturday,True


In [19]:
# dim_marketing — chi phí marketing theo ngày
# Thêm 2 cột phụ trợ: tổng spend và tỉ lệ phân bổ online/offline
dim_marketing = mkt.copy()
dim_marketing['Total_Spend'] = dim_marketing['Offline_Spend'] + dim_marketing['Online_Spend']
dim_marketing['Online_Pct'] = (dim_marketing['Online_Spend']  / dim_marketing['Total_Spend']).round(4)
dim_marketing['Offline_Pct'] = (dim_marketing['Offline_Spend'] / dim_marketing['Total_Spend']).round(4)
dim_marketing.head()

,Date,Offline_Spend,Online_Spend,Total_Spend,Online_Pct,Offline_Pct
0,2019-01-01,4500,2424.50,6924.50,0.3501,0.6499
1,2019-01-02,4500,3480.36,7980.36,0.4361,0.5639
2,2019-01-03,4500,1576.38,6076.38,0.2594,0.7406
3,2019-01-04,4500,2928.55,7428.55,0.3942,0.6058
4,2019-01-05,4500,4055.30,8555.30,0.4740,0.5260


In [20]:
# Xóa các cột đã có trong dim_product, chỉ giữ lại 'Product_SKU' để liên kết
fact_sales = fact_sales.drop(columns=['Product_Description', 'Product_Category','GST'])
fact_sales.head()

,Transaction_ID,CustomerID,Transaction_Date,Year,Quarter,Month,Week,DayOfWeek,Product_SKU,Quantity,Avg_Price,Delivery_Charges,Coupon_Status,Discount_pct,Invoice_Value
0,16679,17850,2019-01-01,2019,1,Jan,1,Tuesday,GGOENEBJ079499,1,153.71,6.5,Used,0.1,158.67
1,16680,17850,2019-01-01,2019,1,Jan,1,Tuesday,GGOENEBJ079499,1,153.71,6.5,Used,0.1,158.67
2,16681,17850,2019-01-01,2019,1,Jan,1,Tuesday,GGOEGFKQ020399,1,2.05,6.5,Used,0.1,8.53
3,16682,17850,2019-01-01,2019,1,Jan,1,Tuesday,GGOEGAAB010516,5,17.53,6.5,Not Used,0.0,109.93
4,16682,17850,2019-01-01,2019,1,Jan,1,Tuesday,GGOEGBJL013999,1,16.50,6.5,Used,0.1,24.02


---
## Tổng kết Star Schema

### fact_sales — Bảng giao dịch chính
52,924 dòng, 15 cột — mỗi dòng là 1 sản phẩm trong 1 đơn hàng.

| Cột | Kiểu | Vai trò | Mô tả |
|---|---|---|---|
| Transaction_ID | int | Mã đơn hàng | 1 đơn có thể gồm nhiều sản phẩm nên cột này bị trùng lặp |
| CustomerID | int | **FK - dim_customer** | ID khách hàng |
| Transaction_Date | datetime | **FK - dim_date** | Ngày giao dịch |
| Year | int | Time feature | Năm giao dịch |
| Quarter | int | Time feature | Quý (1-4) |
| Month | str | Time feature | Tháng viết tắt (Jan, Feb...) |
| Week | int | Time feature | Số tuần trong năm |
| DayOfWeek | str | Time feature | Tên ngày trong tuần |
| Product_SKU | str | **FK - dim_product** | Mã sản phẩm |
| Quantity | int | Số đo | Số lượng mua |
| Avg_Price | float | Số đo | Đơn giá |
| Delivery_Charges | float | Số đo | Phí giao hàng |
| Coupon_Status | str | Thông tin | Used / Clicked / Not Used |
| Discount_pct | float | Số đo | Tỷ lệ giảm giá (0.0 → 1.0) |
| Invoice_Value | float | **Số đo chính** | Tổng giá trị hóa đơn khách hàng thanh toán |

---

### dim_customer — Thông tin khách hàng
1,468 khách hàng unique, 4 cột

| Cột | Kiểu | Vai trò | Mô tả |
|---|---|---|---|
| CustomerID | int | **PK** | ID duy nhất của khách hàng |
| Gender | str | Thuộc tính | Giới tính (M/F) |
| Location | str | Thuộc tính | Khu vực địa lý |
| Tenure_Months | int | Thuộc tính | Số tháng là khách hàng |

---

### dim_product — Danh mục sản phẩm
1,145 SKU unique, 4 cột bao gồm thông tin thuế của từng Product_Category

| Cột | Kiểu | Vai trò | Mô tả |
|---|---|---|---|
| Product_SKU | str | **PK** | Mã sản phẩm duy nhất |
| Product_Description | str | Thuộc tính | Tên, mô tả sản phẩm |
| Product_Category | str | Thuộc tính | Danh mục sản phẩm |
| GST | float | Thuộc tính | Tỷ lệ thuế theo danh mục |

---

### dim_date — Calendar Table
365 ngày liên tục năm 2019.

| Cột | Kiểu | Vai trò | Mô tả |
|---|---|---|---|
| Date | datetime | **PK** | Ngày tháng năm |
| Year | int | Thuộc tính | Năm |
| Quarter | int | Thuộc tính | Quý (1-4) |
| Month_Num | int | Thuộc tính | Số tháng (1-12) |
| Month_Name | str | Thuộc tính | Tên tháng viết tắt |
| Week | int | Thuộc tính | Số tuần ISO |
| DayOfWeek | str | Thuộc tính | Tên ngày trong tuần |
| Is_Weekend | bool | Thuộc tính | True nếu Thứ 7 hoặc Chủ nhật |

---

### dim_marketing — Chi phí Marketing
365 ngày, 6 cột — chi phí marketing offline và online theo ngày.

| Cột | Kiểu | Vai trò | Mô tả |
|---|---|---|---|
| Date | datetime | **PK** | Ngày tháng năm |
| Offline_Spend | int | Số đo | Chi phí TV, radio, báo, billboard |
| Online_Spend | float | Số đo | Chi phí Google Ads, Facebook |
| Total_Spend | float | Số đo | Tổng chi phí marketing |
| Online_Pct | float | Số đo | Tỷ lệ chi phí online trong ngày hôm đó |
| Offline_Pct | float | Số đo | Tỷ lệ chi phí offline trong ngày hôm đó |


---
## Export - lưu các bảng ra file CSV

In [21]:
fact_sales.to_csv('data/processed/fact_sales.csv',index=False)
dim_customer.to_csv('data/processed/dim_customer.csv',index=False)
dim_product.to_csv('data/processed/dim_product.csv',index=False)
dim_marketing.to_csv('data/processed/dim_marketing.csv',index=False)
dim_date.to_csv('data/processed/dim_date.csv',index=False)

---
## Bước tiếp theo : KPI Analysis với SQL

Sau khi hoàn thành xử lý dữ liệu, 5 bảng đã được export ra `data/processed/` 
và import vào **MySQL database `ecommerce_2019`** để phục vụ các phân tích KPI ở phase 2.

In [ ]:
#import các bảng đã xử lí vào mySQL
from sqlalchemy import create_engine

engine = create_engine('mysql+pymysql://root:pass@localhost:3306/ecommerce_2019')

fact_sales.to_sql('fact_sales', con=engine, if_exists='replace', index=False)
dim_customer.to_sql('dim_customer', con=engine, if_exists='replace', index=False)
dim_product.to_sql('dim_product', con=engine, if_exists='replace', index=False)
dim_date.to_sql('dim_date', con=engine, if_exists='replace', index=False)
dim_marketing.to_sql('dim_marketing', con=engine, if_exists='replace', index=False)